public/app/plugins/datasource/loki/formatterTS.ts

```diff
@@ -0,0 +1,406 @@
+import { SyntaxNode } from '@lezer/common';
+import { trimEnd } from 'lodash';
+
+import {
+  Eq,
+  Identifier,
+  String,
+  LogExpr,
+  Matcher,
+  parser,
+  PipelineExpr,
+  Selector,
+  Nre,
+  Re,
+  Neq,
+  LabelParser,
+  LabelFilter,
+  JsonExpressionParser,
+  LineFormatExpr,
+  LabelFormatExpr,
+  LineComment,
+  LineFilter,
+  Filter,
+  Regexp,
+  Pattern,
+  JsonExpression,
+  IpLabelFilter,
+  UnitFilter,
+  NumberFilter,
+  DurationFilter,
+  BytesFilter,
+  Gtr,
+  Gte,
+  Lss,
+  Lte,
+  Eql,
+  Duration,
+  Bytes,
+  Number,
+  LabelFormatMatcher,
+  FilterOp,
+} from '@grafana/lezer-logql';
+
+export const formatLogQL = (query: string): string => {
+  const tree = parser.parse(query);
```

Comment:
To save you headaches, you should run the query through `validateQuery`, and avoid validation if the query is invalid. That would simplify the implementation *a lot*, and prevent unexpected errors.

When that happens, we can display a message to the user, or disable the formatting button.